### PR TITLE
Preprocessor symbols for nullable reference types: Define a single nullable context.

### LINKED ISSUE #1088: [Nullable Reference Types] Compiler directives for nullable reference types
Specify the compiler directive `#nullable` that controls the interpretation of reference types and nullability. The effect of the directives is as follows:

- `#nullable disable`: Sets the nullable annotation and warning contexts to *disabled*.
- `#nullable enable`: Sets the nullable annotation and warning contexts to *enabled*.
- `#nullable restore`: Restores the nullable annotation and warning contexts to project settings.
- `#nullable disable annotations`: Sets the nullable annotation context to *disabled*.
- `#nullable enable annotations`: Sets the nullable annotation context to *enabled*.
- `#nullable restore annotations`: Restores the nullable annotation context to project settings.
- `#nullable disable warnings`: Sets the nullable warning context to *disabled*.
- `#nullable enable warnings`: Sets the nullable warning context to *enabled*.
- `#nullable restore warnings`: Restores the nullable warning context to project settings.

In addition, the spec should define the behavior of reference types in those contexts.

---
[Associated WorkItem - 248771](https://dev.azure.com/msft-skilling/Content/_workitems/edit/248771)

### PR DESCRIPTION
**Pull request by BillWagner**
Fixes #1088 

Alternative PR to #1108

This branch builds on the work in #1108. This is an alternative approach that defines a single `nullable` context. The nullable context has two properties: `annotations` and `warnings` that can be enabled or disabled separately.

### CODE REVIEW

**Review by jskeet — APPROVED** [review]
Yup, personally I like this better than having multiple separate notional contexts.

File: standard/lexical-structure.md
Comment (Nigel-Ecma) [inline standard/lexical-structure.md]: Needs to say both annotations and warnings are enabled.

The previous bullet could also say it disables both, though that is not so important if the  options are:

- null context disabled, or
- null context enabled with annotations and/or warnings

File: standard/lexical-structure.md
Comment (Nigel-Ecma) [inline standard/lexical-structure.md]: `#nullable enable` has an effect if only one of annotations and warnings are currently enabled as it enables the other.

**Review by Nigel-Ecma — CHANGES_REQUESTED** [review]
A couple of language fixes/clarifications required so it describes the same overall semantics as #1108.

**Comment by Nigel-Ecma** [comment]
> Yup, personally I like this better than having multiple separate notional contexts.

I probably agree (yes I'm hedging ;-)) – If a file starts with an enable/disable and never changes it makes little difference. However I can see that when nullable state is enabled & disabled at different points in a file having one context might make that semantic description simpler.

File: standard/lexical-structure.md
Comment (jskeet) [reply standard/lexical-structure.md]: Personally I'd prefer the approach of saying "enables both annotations and warnings".

**Comment by jskeet** [comment]
Example of something that might be hard to understand:

```csharp
#nullable disable
string x = null;
string y = "";
#nullable enable
Console.WriteLine(x.ToString());
Console.WriteLine(y.ToString());
```

Should that give a warning for `x.ToString()`?

#1124 states: "However, the disabled context means the compiler does not track null state and does not issue nullable related warnings" - so that would suggest no warning.

However, (with .NET 8 SDK) it issues a warning:

warning CS8602: Dereference of a possibly null reference.

File: standard/lexical-structure.md
Comment (jskeet) [reply standard/lexical-structure.md]: In meeting: no need for this.

File: standard/lexical-structure.md
Comment (jskeet) [reply standard/lexical-structure.md]: In meeting: yes, make each of these specify the effect on both flags.

**Review by jskeet — APPROVED** [review]
Approved but with changes as described.

**Comment by BillWagner** [comment]
I updated everything from the last meeting (both comments, and from the in-meeting discussion). This should be ready to merge @jskeet 

We can discuss at the meeting, or merge before.

**Comment by jskeet** [comment]
> We can discuss at the meeting, or merge before.

Let's get a final review at the meeting just in case. It would be handy if everyone could review beforehand though.

File: standard/lexical-structure.md
Comment (Nigel-Ecma) [inline standard/lexical-structure.md:1029]: Reading this now “nullable context” caused a pause. We have a null analysis, which may be enabled or disabled; and certain annotations (like adding a `?` to a type) etc. which both provide input to that analysis and which are treated differently depending on whether the analysis is being done. Is this a “nullable context”, a ”null analysis context”, something else?

File: standard/lexical-structure.md
Comment (Nigel-Ecma) [inline standard/lexical-structure.md]: “until the end of the compilation unit is reached”? (or *compilation_unit*)

File: standard/lexical-structure.md
Comment (Nigel-Ecma) [inline standard/lexical-structure.md]: “The nullable state of expressions…” vs. “…initial null state of a variable…”
Change the former to “null”?
“is tracked” -> ”shall be tracked” – that’s one for @RexJaeschke to determine, but the tense seems wrong to me and maybe the Standardese.

File: standard/lexical-structure.md
Comment (Nigel-Ecma) [inline standard/lexical-structure.md]: I would replace:

> (“dereference of a possible null reference”)

with

> (as `x` is `null`)

as the former reads like it is specifying the error text, including “possible”, even though we know its non-normative.

Of course this also states *non normatively* that there will be a warning, we is also not currently required…

File: standard/lexical-structure.md
Comment (gafter) [inline standard/lexical-structure.md]: The state of the annotation flag *and* the presence or absence of a nullable annotation `?`.

**Review by Nigel-Ecma — DISMISSED** [review]
A few language issues.

I also do not think we should approve/merge this, it is one of a set and as the others are developed we might need to circle around to this or any of them. So we should leave it as "done for now", a well known Github state ;-)

File: standard/lexical-structure.md
Comment (BillWagner) [inline standard/lexical-structure.md]:
```suggestion
The nullable state of expressions is tracked at all times. The state of the annotation flag and the presence or absence of a nullable annotation, `?`, determines the initial null state of a variable declaration. Warnings are only issued when the warnings flag is enabled.
```

File: standard/lexical-structure.md
Comment (BillWagner) [inline standard/lexical-structure.md]:
```suggestion
> produces a compile-time warning (“as `x` is `null`”). The nullable state of `x` is tracked everywhere. A warning is issued when the warnings flag is enabled.
```

File: standard/lexical-structure.md
Comment (BillWagner) [reply standard/lexical-structure.md:1029]: Tracked in the epic.

File: standard/lexical-structure.md
Comment (BillWagner) [inline standard/lexical-structure.md]:
```suggestion
A nullable directive sets the available flags for subsequent lines of code, until another nullable directive overrides it, or until the end of the *compilation _unit* is reached. The nullable context contains two flags: *annotations* and *warnings*. The effect of each form of nullable directive is, as follows:
```

**Review by jskeet — APPROVED** [review]

**Comment by jskeet** [comment]
> I also do not think we should approve/merge this, it is one of a set and as the others are developed we might need to circle around to this or any of them. So we should leave it as "done for now", a well known Github state ;-)

This was discussed unfortunately in Nigel's absence due to connectivity issues. We agreed that it would be better to avoid having multiple PRs hanging around and getting stale - the spec is "always inconsistent" in a draft form, as it doesn't accurately describe any specific version of C#.

If Nigel wishes to convince us otherwise, we could always revert the commit and open another PR :)

We will dismiss Nigel's review in order to merge now.

File: standard/lexical-structure.md
Comment (Nigel-Ecma) [reply standard/lexical-structure.md:1029]: Re: [epic comment](https://github.com/dotnet/csharpstandard/issues/1085#issuecomment-2221397275) – yes “is being done” should be worded ”is enabled/disabled”. Which unfortunately highlights something we are stuck with – the analysis is always being done but the pragmas enable/disable it… C’est la vie

**Pull request merged**
